app/controllers/claimToAdjustPoa/SelectYourReasonController.scala

```diff
@@ -0,0 +1,157 @@
+                                            val authorisedFunctions: AuthorisedFunctions,
+                                            implicit val itvcErrorHandler: ItvcErrorHandler,
+                                            implicit val itvcErrorHandlerAgent: AgentItvcErrorHandler,
+                                            auth: AuthenticatorPredicate)
+                                         (implicit val appConfig: FrontendAppConfig,
+                                          implicit override val mcc: MessagesControllerComponents,
+                                          val ec: ExecutionContext)
+  extends ClientConfirmedController with I18nSupport with FeatureSwitching  {
+
+  def show(isAgent: Boolean, isChange: Boolean): Action[AnyContent] = auth.authenticatedAction(isAgent = isAgent) {
+    implicit user =>
+      ifAdjustPoaIsEnabled(isAgent) {
+        withSessionAndPoa(isAgent) { (session, poa) =>
+          (session.newPoAAmount) match {
+            case Some(amount) if amount >= poa.totalAmount =>
+              saveValueAndRedirect(isChange, isAgent, Increase)
+            case _ =>
+              val form = formProvider.apply()
+              EitherT.rightT(Ok(view(
+                selectYourReasonForm = session.poaAdjustmentReason.fold(form)(form.fill),
+                taxYear = poa.taxYear,
+                isAgent = isAgent,
+                isChange = isChange,
+                useFallbackLink = true)))
+          }
+        } fold(
+          logAndShowErrorPage(isAgent, "show"),
+          view => view
+        )
+      }
+  }
+
+  def submit(isAgent: Boolean, isChange: Boolean): Action[AnyContent] = auth.authenticatedAction(isAgent = isAgent) {
+    implicit request =>
+      ifAdjustPoaIsEnabled(isAgent) {
+        formProvider.apply()
+          .bindFromRequest()
+          .fold(
+            formWithErrors => withSessionAndPoa(isAgent) { (_, poa) =>
+              EitherT.rightT(BadRequest(view(formWithErrors, poa.taxYear, isAgent, isChange, true)))
+            },
+            value => saveValueAndRedirect(isChange, isAgent, value))
+          .fold(
+            logAndShowErrorPage(isAgent, "submit"),
+            result => result
+          )
+      }
+  }
+
+  private def withSessionAndPoa (isAgent: Boolean)
+                                (block: (PoAAmendmentData, PaymentOnAccountViewModel) => EitherT[Future, Throwable, Result])
+                                (implicit user: MtdItUser[_]): EitherT[Future, Throwable, Result] = {
+
+    val errorHandler = if (isAgent) itvcErrorHandlerAgent else itvcErrorHandler
+
+    for {
+      session <- EitherT(sessionService.getMongo)
+      poa <- EitherT(claimToAdjustService.getPoaForNonCrystallisedTaxYear(Nino(user.nino)))
+      result <- (session, poa) match {
+        case (Some(s), Some(p)) =>
+          block(s, p)
+        case (None, _) =>
+          Logger("application").error(s"[SelectYourReasonController][withSessionAndPoa] session missing")
```

Comment:
no need to specify source of the problem in the log entry, I think we recently merge this change and it manager by logback configuration, so you can drop prefix like: "[SelectYourReasonController][withSessionAndPoa].."